rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -902,3 +1121,154 @@ impl<Tokens: TokensType> Ledger<Tokens> {
+            .map(|a| a.into())
+    }
+
+    fn set_allowance(
+        &mut self,
+        account_spender: (Self::AccountId, Self::AccountId),
+        allowance: Allowance<Self::Tokens>,
+    ) {
+        let account_spender = (&account_spender).into();
+        ALLOWANCES_MEMORY
+            .with_borrow_mut(|allowances| allowances.insert(account_spender, allowance.into()));
+    }
+
+    fn remove_allowance(&mut self, account_spender: &(Self::AccountId, Self::AccountId)) {
+        let account_spender = account_spender.into();
+        ALLOWANCES_MEMORY.with_borrow_mut(|allowances| allowances.remove(&account_spender));
+    }
+
+    fn insert_expiry(
+        &mut self,
+        timestamp: TimeStamp,
+        account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        let account_spender = (&account_spender).into();
+        let expiration = Expiration {
+            timestamp,
+            account_spender,
+        };
+        ALLOWANCES_EXPIRATIONS_MEMORY.with_borrow_mut(|expirations| {
+            expirations.insert(expiration, ());
+        });
+    }
+
+    fn remove_expiry(
+        &mut self,
+        timestamp: TimeStamp,
+        account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        let account_spender = (&account_spender).into();
+        let expiration = Expiration {
+            timestamp,
+            account_spender,
+        };
+        ALLOWANCES_EXPIRATIONS_MEMORY.with_borrow_mut(|expirations| {
+            expirations.remove(&expiration);
+        });
+    }
+
+    fn insert_arrival(
+        &mut self,
+        _timestamp: TimeStamp,
+        _account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        // We do not store arrivals in stable structures.
+    }
+
+    fn remove_arrival(
+        &mut self,
+        _timestamp: TimeStamp,
+        _account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        // We do not store arrivals in stable structures.
+    }
+
+    fn first_expiry(&self) -> Option<(TimeStamp, (Self::AccountId, Self::AccountId))> {
+        let result = ALLOWANCES_EXPIRATIONS_MEMORY
+            .with_borrow(|expirations| expirations.first_key_value().map(|kv| kv.0));
+        result.map(|e| (e.timestamp, e.account_spender.into()))
+    }
+
+    fn oldest_arrivals(&self, _n: usize) -> Vec<(Self::AccountId, Self::AccountId)> {
+        // We do not store arrivals in stable structures.
+        vec![]
+    }
+
+    fn pop_first_expiry(&mut self) -> Option<(TimeStamp, (Self::AccountId, Self::AccountId))> {
+        let result = ALLOWANCES_EXPIRATIONS_MEMORY
+            .with_borrow_mut(|expirations| expirations.pop_first().map(|kv| kv.0));
+        result.map(|e| (e.timestamp, e.account_spender.into()))
+    }
+
+    fn pop_first_allowance(
+        &mut self,
+    ) -> Option<((Self::AccountId, Self::AccountId), Allowance<Self::Tokens>)> {
+        panic!("The method `pop_first_allowance` should not be called for StableAllowancesData")
```

Comment:
Are we sure this won't be called or can this panic be hit on some code path?

Reply:
It is only called in one place, for the old approvals `self.approvals.allowances_data.pop_first_allowance()`

Reply:
So it is never called for stable approvals?

Reply:
Correct!